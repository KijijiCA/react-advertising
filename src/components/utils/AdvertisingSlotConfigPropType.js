import PropTypes from 'prop-types';

export default PropTypes.shape({
    id: PropTypes.string.isRequired,
    path: PropTypes.string,
    collapseEmptyDiv: PropTypes.arrayOf(PropTypes.bool),
    targeting: PropTypes.object,
    sizes: PropTypes.oneOfType([
        PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.number)]),
        PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.number)])),
    ]),
    sizeMappingName: PropTypes.string,
    prebid: PropTypes.arrayOf(
        PropTypes.shape({
            mediaTypes: PropTypes.objectOf(
                PropTypes.shape({
                    sizes: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
                })
            ).isRequired,
            bids: PropTypes.arrayOf(
                PropTypes.shape({
                    bidder: PropTypes.string.isRequired,
                    params: PropTypes.object,
                    labelAny: PropTypes.arrayOf(PropTypes.string),
                    labelAll: PropTypes.arrayOf(PropTypes.string),
                })
            ).isRequired,
        })
    ).isRequired,
});
