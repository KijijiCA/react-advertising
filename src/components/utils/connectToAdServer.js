import React from 'react';
import AdvertisingContext from '../../AdvertisingContext';

export default (Component) => (props) => (
    <AdvertisingContext.Consumer>
        {(activate) => <Component {...props} activate={activate} />}
    </AdvertisingContext.Consumer>
);
