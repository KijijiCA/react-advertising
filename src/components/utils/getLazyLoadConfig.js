export default function getLazyLoadConfig(config, id) {
  if (!config?.slots) {
    return null;
  }
  const slotConfig = config.slots.find((slot) => slot.id === id);
  if (!slotConfig) {
    return null;
  }
  if (
    slotConfig.enableLazyLoad !== undefined &&
    slotConfig.enableLazyLoad !== null
  ) {
    return slotConfig.enableLazyLoad;
  }
  return config?.enableLazyLoad === undefined ? null : config?.enableLazyLoad;
}
