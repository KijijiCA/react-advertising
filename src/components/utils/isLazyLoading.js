export default function isLazyLoading(lazyLoadConfig) {
  if (typeof lazyLoadConfig === 'boolean') {
    return lazyLoadConfig;
  }

  if (!lazyLoadConfig) {
    return false;
  }

  const { marginPercent } = lazyLoadConfig;

  if (!marginPercent) {
    return true;
  }

  return marginPercent > -1;
}
