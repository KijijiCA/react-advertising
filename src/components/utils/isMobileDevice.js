// Mobile device detection, based on MDN solution:
// https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent
export default function isMobileDevice() {
  if ('maxTouchPoints' in navigator) {
    return navigator.maxTouchPoints > 0;
  }
  if ('msMaxTouchPoints' in navigator) {
    return navigator.msMaxTouchPoints > 0;
  }
  const mQ = window.matchMedia && matchMedia('(pointer:coarse)');
  if (mQ && mQ.media === '(pointer:coarse)') {
    return !!mQ.matches;
  }
  if ('orientation' in window) {
    return true; // deprecated, but good fallback
  }
  // Only as a last resort, fall back to user agent sniffing
  const UA = navigator.userAgent;
  return (
    /\b(BlackBerry|webOS|iPhone|IEMobile)\b/i.test(UA) ||
    /\b(Android|Windows Phone|iPad|iPod)\b/i.test(UA)
  );
}
