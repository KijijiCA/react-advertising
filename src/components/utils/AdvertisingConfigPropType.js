import PropTypes from 'prop-types';
import AdvertisingSlotConfigPropType from './AdvertisingSlotConfigPropType';

export default PropTypes.shape({
  path: PropTypes.string,
  targeting: PropTypes.object,
  usePrebid: PropTypes.bool,
  prebid: PropTypes.shape({
    debug: PropTypes.bool,
    bidderTimeout: PropTypes.number,
    enableSendAllBids: PropTypes.bool,
    bidderSequence: PropTypes.oneOf(['random', 'fixed']),
    publisherDomain: PropTypes.string,
    cookieSyncDelay: PropTypes.number,
    priceGranularity: PropTypes.oneOfType([
      PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
      PropTypes.shape({
        buckets: PropTypes.arrayOf(
          PropTypes.shape({
            precision: PropTypes.number,
            min: PropTypes.number.isRequired,
            max: PropTypes.number.isRequired,
            increment: PropTypes.number.isRequired,
          })
        ).isRequired,
      }),
    ]),
    mediaTypePriceGranularity: PropTypes.shape({
      video: PropTypes.oneOfType([
        PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
        PropTypes.shape({
          buckets: PropTypes.arrayOf(
            PropTypes.shape({
              precision: PropTypes.number,
              min: PropTypes.number.isRequired,
              max: PropTypes.number.isRequired,
              increment: PropTypes.number.isRequired,
            })
          ).isRequired,
        }),
      ]),
      banner: PropTypes.oneOfType([
        PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
        PropTypes.shape({
          buckets: PropTypes.arrayOf(
            PropTypes.shape({
              precision: PropTypes.number,
              min: PropTypes.number.isRequired,
              max: PropTypes.number.isRequired,
              increment: PropTypes.number.isRequired,
            })
          ).isRequired,
        }),
      ]),
      native: PropTypes.oneOfType([
        PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
        PropTypes.shape({
          buckets: PropTypes.arrayOf(
            PropTypes.shape({
              precision: PropTypes.number,
              min: PropTypes.number.isRequired,
              max: PropTypes.number.isRequired,
              increment: PropTypes.number.isRequired,
            })
          ).isRequired,
        }),
      ]),
    }),
    sizeConfig: PropTypes.arrayOf(
      PropTypes.shape({
        mediaQuery: PropTypes.string.isRequired,
        sizesSupported: PropTypes.arrayOf(
          PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.arrayOf(PropTypes.number),
          ])
        ),
        labels: PropTypes.arrayOf(PropTypes.string),
      })
    ),
  }),
  sizeMappings: PropTypes.objectOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        viewPortSize: PropTypes.arrayOf(PropTypes.number).isRequired,
        sizes: PropTypes.arrayOf(
          PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.arrayOf(PropTypes.number),
          ])
        ).isRequired,
      })
    )
  ),
  slots: PropTypes.arrayOf(AdvertisingSlotConfigPropType),
  outOfPageSlots: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
    })
  ),
  customEvents: PropTypes.objectOf(
    PropTypes.shape({
      eventMessagePrefix: PropTypes.string.isRequired,
      divIdPrefix: PropTypes.string,
    })
  ),
  enableLazyLoad: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.shape({
      marginPercent: PropTypes.number,
      mobileScaling: PropTypes.number,
    }),
  ]),
});
