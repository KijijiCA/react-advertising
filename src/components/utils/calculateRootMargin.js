import isMobileDevice from './isMobileDevice';

export default function calculateRootMargin({
  marginPercent,
  mobileScaling,
} = {}) {
  if (!marginPercent) {
    return undefined;
  }
  const finalMarginPercent =
    isMobileDevice() && mobileScaling !== undefined && mobileScaling !== -1
      ? marginPercent * mobileScaling
      : marginPercent;
  return `${(finalMarginPercent / 100) * window.innerHeight}px`;
}
