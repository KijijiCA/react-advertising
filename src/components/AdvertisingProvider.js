import React, { Component } from 'react';
import Advertising from '../Advertising';
import PropTypes from 'prop-types';
import AdvertisingConfigPropType from './utils/AdvertisingConfigPropType';
import AdvertisingContext from '../AdvertisingContext';
import equal from 'fast-deep-equal';

export default class AdvertisingProvider extends Component {
  constructor(props) {
    super(props);
    this.initialize();

    this.state = {
      activate: this.advertising.activate.bind(this.advertising),
    };
  }

  async componentDidMount() {
    if (this.advertising.isConfigReady() && this.props.active) {
      await this.advertising.setup();
    }
  }

  async componentDidUpdate(prevProps) {
    // this means teardown method has been invoked already
    if (!this.advertising) {
      return;
    }

    const { config, active } = this.props;
    const isConfigReady = this.advertising.isConfigReady();

    // activate advertising when the config changes from `undefined`
    if (!isConfigReady && config && active) {
      this.advertising.setConfig(config);
      await this.advertising.setup();
    } else if (isConfigReady && !equal(prevProps.config, config)) {
      // teardown the old configuration
      // to make sure the teardown and initialization are in a right sequence, need `await`
      await this.teardown();

      // re-initialize advertising, if it is active
      if (active) {
        this.initialize();
        // eslint-disable-next-line react/no-did-update-set-state
        this.setState({
          activate: this.advertising.activate.bind(this.advertising),
        });

        if (this.advertising.isConfigReady()) {
          await this.advertising.setup();
        }
      }
    }
  }

  async componentWillUnmount() {
    if (this.advertising && this.props.config) {
      await this.teardown();
    }
  }

  async teardown() {
    await this.advertising.teardown();
    this.advertising = null;
    this.activate = null;
  }

  initialize() {
    const { config, plugins, onError } = this.props;
    this.advertising = new Advertising(config, plugins, onError);
  }

  render() {
    const { activate } = this.state;
    return (
      <AdvertisingContext.Provider value={activate}>
        {this.props.children}
      </AdvertisingContext.Provider>
    );
  }
}

AdvertisingProvider.propTypes = {
  active: PropTypes.bool,
  config: AdvertisingConfigPropType,
  children: PropTypes.node,
  onError: PropTypes.func,
  plugins: PropTypes.arrayOf(
    PropTypes.shape({
      setupPrebid: PropTypes.func,
      setupGpt: PropTypes.func,
      teardownPrebid: PropTypes.func,
      teardownGpt: PropTypes.func,
    })
  ),
};

AdvertisingProvider.defaultProps = {
  active: true,
};
