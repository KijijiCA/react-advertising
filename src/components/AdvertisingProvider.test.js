import React from 'react';
import { render } from '@testing-library/react';
import AdvertisingProvider from './AdvertisingProvider';
import { config } from '../utils/testAdvertisingConfig';

const mockActivate = jest.fn();
const mockSetup = jest.fn();
const mockTeardown = jest.fn();
const mockConstructor = jest.fn();
const mockValueSpy = jest.fn();
const mockIsConfigReady = jest.fn();
const mockSetConfig = jest.fn();

jest.mock('../AdvertisingContext', () => ({
  // eslint-disable-next-line react/prop-types
  Provider({ value, children }) {
    mockValueSpy(value);
    return <div id="advertising-context-provider">{children}</div>;
  },
}));

jest.mock(
  '../Advertising',
  () =>
    class {
      constructor(...args) {
        mockConstructor(...args);
      }
      activate(...args) {
        mockActivate(...args);
      }
      async setup(...args) {
        mockSetup(...args);
      }
      async teardown(...args) {
        mockTeardown(...args);
      }
      isConfigReady() {
        return mockIsConfigReady();
      }
      setConfig(...args) {
        mockSetConfig(...args);
      }
    }
);

/* eslint-disable no-unused-expressions */
describe('The AdvertisingProvider component', () => {
  describe('when provided a configuration from the start', () => {
    let provider, rerender, unmount;

    beforeEach(() => {
      jest.clearAllMocks();
      mockIsConfigReady.mockReturnValue(true);
      ({
        container: { firstChild: provider },
        rerender,
        unmount,
      } = render(
        <AdvertisingProvider config={config}>
          <h1>hello</h1>
        </AdvertisingProvider>
      ));
    });

    it('renders correctly', () => {
      expect(provider).toMatchSnapshot();
    });

    it('constructs an Advertising module with the provided configuration', () => {
      expect(mockConstructor).toHaveBeenCalledWith(
        config,
        undefined,
        undefined
      );
    });

    it('sets up the Advertising module', () => {
      expect(mockSetup).toHaveBeenCalled();
    });

    it('calls setup only once even if config reference is changed but the content has not changed', () => {
      rerender(<AdvertisingProvider config={{ ...config }} />);
      expect(mockSetup).toHaveBeenCalledTimes(1);
    });

    it('calls setup when the config content is changed', (done) => {
      rerender(
        <AdvertisingProvider
          config={{ ...config, path: 'global/ad/unit/path2' }}
        />
      );

      setTimeout(() => {
        expect(mockSetup).toHaveBeenCalledTimes(2);
        expect(mockTeardown).toHaveBeenCalledTimes(1);
        done();
      }, 0);
    });

    it('creates a new Advertising instance with a new config when the config is changed', (done) => {
      rerender(<AdvertisingProvider />);

      setTimeout(() => {
        expect(mockConstructor).toHaveBeenCalledWith(
          undefined,
          undefined,
          undefined
        );
        expect(mockTeardown).toHaveBeenCalledTimes(1);
        done();
      }, 0);
    });

    it('does not call setup if the config content is changed but active is `false`', (done) => {
      rerender(
        <AdvertisingProvider
          config={{ ...config, path: 'global/ad/unit/path2' }}
          active={false}
        />
      );

      setTimeout(() => {
        expect(mockSetup).toHaveBeenCalledTimes(1);
        done();
      }, 0);
    });

    it('does not call setup if the config changes to `undefined`', (done) => {
      mockIsConfigReady.mockReturnValueOnce(false);
      rerender(<AdvertisingProvider />);

      setTimeout(() => {
        expect(mockTeardown).toHaveBeenCalledTimes(0);
        expect(mockSetup).toHaveBeenCalledTimes(1);
        done();
      }, 0);
    });

    it('uses an AdvertisingContext.Provider to pass the activate method of the advertising module', () => {
      expect(mockValueSpy.mock.calls[0][0]).toMatchSnapshot();
    });

    it('tears down the Advertising module when it unmounts', () => {
      unmount();
      expect(mockTeardown).toHaveBeenCalledTimes(1);
    });
  });

  describe('when config is loaded asynchronously and provided later', () => {
    let rerender, unmount;
    beforeEach(() => {
      jest.clearAllMocks();
      mockIsConfigReady.mockReturnValue(false);
      ({ rerender, unmount } = render(<AdvertisingProvider />));
    });

    it('constructs an Advertising module', () => {
      expect(mockConstructor).toHaveBeenCalled();
    });

    it('sets configuration and sets up the Advertising module', () => {
      rerender(<AdvertisingProvider config={config} />);

      expect(mockSetConfig).toHaveBeenCalledWith(config);
      expect(mockSetup).toHaveBeenCalledTimes(1);
    });

    it('tears down the Advertising module', (done) => {
      rerender(<AdvertisingProvider config={config} />);
      unmount();

      setTimeout(() => {
        expect(mockTeardown).toHaveBeenCalledTimes(1);
        done();
      }, 0);
    });

    it('does not tear down the Advertising module if the config is undefined', () => {
      unmount();

      expect(mockTeardown).toHaveBeenCalledTimes(0);
    });
  });

  describe('when mounted with active = false', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      mockIsConfigReady.mockReturnValue(true);
      render(<AdvertisingProvider config={config} active={false} />);
    });

    it('constructs an Advertising module with the provided configuration', () => {
      expect(mockConstructor).toHaveBeenCalledWith(
        config,
        undefined,
        undefined
      );
    });

    it('does not set up the Advertising module', () => {
      expect(mockSetup).toHaveBeenCalledTimes(0);
    });
  });
});
