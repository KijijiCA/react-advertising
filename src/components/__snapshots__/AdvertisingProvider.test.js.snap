// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AdvertisingProvider component when provided a configuration from the start renders correctly 1`] = `
<div
  id="advertising-context-provider"
>
  <h1>
    hello
  </h1>
</div>
`;

exports[`The AdvertisingProvider component when provided a configuration from the start uses an AdvertisingContext.Provider to pass the activate method of the advertising module 1`] = `[Function]`;
